src/overcooked_ai_py/agents/benchmarking.py

```diff
@@ -19,36 +20,76 @@ class AgentEvaluator(object):
+            mdp_fn_naive = LayoutGenerator.mdp_gen_fn_from_dict(mdp_params, outer_shape, mdp_params_schedule_fn)
             # infinite mdp
-            if 'num_mdp' not in env_params or env_params['num_mdp'] == -1:
-                self.mdp_fn = mdp_fn
-                self.env = OvercookedEnv(self.mdp_fn, **env_params)
-            else:
+            if 'num_mdp' not in env_params or np.isinf(env_params['num_mdp']):
+                mdp_fn = mdp_fn_naive
+            else: # finite mdp, random choice
                 num_mdp = env_params['num_mdp']
-                assert num_mdp > 0, "invalid number of mdp"
-                self.mdp_lst = [mdp_fn() for _ in range(num_mdp)]
-                self.mdp_fn = lambda : np.random.choice(self.mdp_lst)
-                self.env = OvercookedEnv(self.mdp_fn, **env_params)
+                assert type(num_mdp) == int and num_mdp > 0, "invalid number of mdp: " + str(num_mdp)
+                mdp_lst = [mdp_fn_naive() for _ in range(num_mdp)]
+                mdp_fn = lambda : np.random.choice(mdp_lst)
+        if "outer_shape" in env_params:
+            del env_params["outer_shape"]
+        return AgentEvaluator(env_params, mdp_fn, force_compute, mlam_params, debug)
 
-        self.force_compute = force_compute
+    @staticmethod
+    def from_mdp(mdp, env_params, force_compute=False, mlam_params=NO_COUNTERS_PARAMS, debug=False):
+        """
+        mdp (OvercookedGridworld): the mdp that we want the AgentEvaluator to always generate
+        Information for the rest of params please refer to the __init__ method above
+        """
+        assert type(mdp) == OvercookedGridworld, "mdp must be a OvercookedGridworld object"
+        mdp_fn = lambda: mdp
+        return AgentEvaluator(env_params, mdp_fn, force_compute, mlam_params, debug)
+
+    @staticmethod
+    def from_mdp_lst(mdp_lst, env_params, sampling_freq=None, force_compute=False, mlam_params=NO_COUNTERS_PARAMS, debug=False):
+        """
+        mdp_lst (list): a list of mdp (OvercookedGridworld) we would like to
+        sampling_freq (list): a list of number that signify the sampling frequency of each mdp in the mdp_lst
+        Information for the rest of params please refer to the __init__ method above
+        """
+        assert type(mdp_lst) == type([]), "mdp must be a OvercookedGridworld object"
+        for mdp in mdp_lst:
+            assert type(mdp) == OvercookedGridworld, "some mdp in the mdp_lst are not OvercookedGridworld"
+
+        if sampling_freq is None:
+            mdp_fn = lambda : np.random.choice(mdp_lst)
+        else:
+            assert len(sampling_freq) == len(mdp_lst), "sampling frequency is not the same length as the mdp_lst"
+            assert np.sum(sampling_freq) == 1.0
+            mdp_fn = lambda : np.random.choice(mdp_lst, p=sampling_freq)
+        return AgentEvaluator(env_params, mdp_fn, force_compute, mlam_params, debug)
```

Comment:
Three staticmethod has been added, the `from_mdp_lst` gives the the possibility to do any arbitrary sampling frequency

Reply:
I agree with what you said, and I think it makes sense.

The current version of __init__ is able to accommodate the need you describe (crazy sampling scheme) by taking in the `mdp_fn` only. Could you let me know if this was not the desired behavior?

The staticmethod was introduced to save work if we don't want to hand write custom sampling scheme. Instead, we just pass in a mdp_lst and a sampling frequency and have the staticmethod do the dirty job. That was why I included the possibility to do a custom sampling frequency.

I will give the Clean Coding series a lot.

Reply:
Yeah that makes sense – I guess I was wondering whether you actually expect to use a custom sampling scheme, rather than just sample uniformly. If not, we could remove the custom `sampling_freq`, and if somebody really wants to do that, they can just use the `__init__` method using `mdp_fn`. Alternatively you could keep the `sampling_freq` param, and clean the code this way

```
if sampling_freq is None:
   sampling_freq = np.ones(len(mdp_lst)) / len(mdp_lst)

mdp_fn = lambda : np.random.choice(mdp_lst, p=sampling_freq)
```

Reply:
Another code improvement here is:

```
assert all(type(mdp) == OvercookedGridworld for mdp in mdp_lst), "some mdp in the mdp_lst are not
```

Reply:
Addressed, marking as resolved